jfrog-oauth/main.test.ts

```diff
@@ -1,19 +1,115 @@
-import { serve } from "bun";
-import { describe } from "bun:test";
+import { describe, expect, it } from "bun:test";
 import {
-  createJSONResponse,
+  findResourceInstance,
   runTerraformInit,
+  runTerraformApply,
   testRequiredVariables,
 } from "../test";
 
 describe("jfrog-oauth", async () => {
   await runTerraformInit(import.meta.dir);
 
-  testRequiredVariables(import.meta.dir, {
-    agent_id: "some-agent-id",
-    jfrog_url: "http://localhost:8081",
-    package_managers: "{}",
+  const fakeFrogHostAndPort = "localhost:8081";
+  const fakeFrogUrl = `http://${fakeFrogHostAndPort}`;
+
+  it("can run apply with required variables", async () => {
+    testRequiredVariables(import.meta.dir, {
```

Comment:
Just to provide a little more type-safety, could you define a custom type for the variables, and pass it along as a type parameter?
    
```ts
type TestVariables = Readonly<{
  agent_id: string;
  jfrog_url: string;
  package_managers: string;

  username_field?: string; 
  jfrog_server_id?: string;
  external_auth_id: string;
  configure_code_server?: boolean;
}>;
```
You'd then be able to pass this type to both `runTerraformApply` and `testRequiredVariables` to catch any accidental typos:
```ts
runTerraformApply<TestVariables>(/* Same arguments as before */)
```

Reply:
I appreciate the concept here but please allow me to push back slightly on this suggestion. I fear it might be redundant to define the variables in the test with minimal value because:

1. This is not at all in the runtime path, we're just potentially saving a developer testing time
2. Typos will bubble up as a result of running the mock terraform apply and the tests will fail
3. Since there is no definitive binding between the variables as written in tf to the variables in this type def, there's nothing to prevent someone from typing the variable name incorrectly in the typescript type def.

In summary, I think you are only validating the test code is type safe and not really validating the logic of the module.

I don't see this pattern in many, if any, of the other tests for modules but if you are adamant that this is the correct thing for your code, I will implement it.

Reply:
I put this in place in the interest of getting these changes merged. I do think it feels contrived since you have to define the type and then make sure you specify the type as a type param for the various function calls to get the benefit of enforcing the type check.

It might be a tighter workflow if you had a generic Testing class instead of the free functions in `test.ts`. So you would do something like:

```js
describe("some-module", async () => {
  type SomeVariables = {
    ...
  };
  const test = new Testing<SomeVariables>();
  const state = await test.runTerraformApply(...);
  ...
});
```

Obviously a change outside the scope of this PR, but does that make sense to you?

Reply:
Sorry, was a little swamped these past few days, and haven't had a chance to respond earlier

On one hand, I guess I view the current setup as not that different from other JavaScript values like Maps or Sets. If you have something like this:
```ts
const userMap = Map<string, User>();
```
I don't think that's too different from one of the testing functions. The type restrictions don't exist at runtime, so non-user values can still accidentally be placed in the `Map`.

But I think you're right, and I think the current type behavior is a bit of a bandaid. These _are_ just test files, so them breaking on typos isn't as huge of a deal, because you get instant feedback before they bubble up to an end-user

But going to your idea about the generic testing class, that does seem like a really good idea. The reason why having `Map<string, User>` is valuable is because the type restriction trickles down to each individual method. You don't have to keep repeating the type info on every single call

Reply:
I think I'll actually make a separate PR once this one is done to leverage classes more